![logo-gt.png](attachment:logo-gt.png)

<div class="alert alert-block alert-info">
    <h1>BIOL 4150/BIOL 6150</h1>
    <h3>Instructor: Dr. King Jordan</h3>
    <p>Shivam Sharma (shivamsharma13@gatech.edu), Nilavrah Sensarma (nsensarma3@gatech.edu), Bengy Gyimah-Asamoah, (bgyimaha3@gatech.edu)

</p>
</div>

<div class="alert alert-block alert-warning">
    <h2>Project 2 (Read Mapping) starter notebook</h2>
    <h3>Deadline: 11:59PM, October 1st, 2024</h3>
</div>

---

# **1. Getting ready for read alignment**
### *Total Questions: 5*
### *Total Points: 2+5+5+5+3 = 20*


---

<div class="alert alert-block alert-warning">
    <h3>1.1 Do you have the post-QCed fastq file from the assigned 1000 Genomes individual</h3>
    <p>We will start from where we left off in Project #1. We want to make sure that we are using the correct QCed fastq files.</p>
</div>

In [4]:
##Run "ls -lh" here to show the two fastq files.
!ls -lh ~/biol6150/ProjectSubmissions/Group8/Project1/fastQCAfterTrimming

#Question: Does showing the HTML files matter?

#/SRR393028_1.Trimmed_fastqc.zip
#!ls -lh ~/biol6150/ProjectSubmissions/Group8/Project1/fastQCAfterTrimming/SRR393028_2.Trimmed_fastqc.zip

total 1.7M
-rw-r--r--. 1 soliver35 gtperson 583K Sep 21 17:22 SRR393028_1.Trimmed_fastqc.html
-rw-r--r--. 1 soliver35 gtperson 262K Sep 21 17:22 SRR393028_1.Trimmed_fastqc.zip
-rw-r--r--. 1 soliver35 gtperson 583K Sep 21 17:22 SRR393028_2.Trimmed_fastqc.html
-rw-r--r--. 1 soliver35 gtperson 261K Sep 21 17:22 SRR393028_2.Trimmed_fastqc.zip


<div class="alert alert-block alert-warning">
    <h3>1.2 Download the correct reference genome for alignment step</h3>
    <p>Please locate and donwload the human reference genome fasta file in build GRCh38</p>
</div>

In [4]:
#Where does this file come from? Give the website link (FTP?) or explain how you downloaed this file.
#Also run "ls -lh" here to see the size and location of your reference file.

In [2]:
!wget http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
!ls -lh

--2024-09-21 18:21:22--  http://hgdownload.soe.ucsc.edu/goldenPath/hg38/bigZips/hg38.fa.gz
Resolving hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)... 128.114.119.163
Connecting to hgdownload.soe.ucsc.edu (hgdownload.soe.ucsc.edu)|128.114.119.163|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 983659424 (938M) [application/x-gzip]
Saving to: ‘hg38.fa.gz’

hg38.fa.gz          100%[===================>] 938.09M  14.1MB/s    in 53s     

2024-09-21 18:22:15 (17.8 MB/s) - ‘hg38.fa.gz’ saved [983659424/983659424]

total 942M
-rw-r--r--. 1 soliver35 gtperson 939M Jan 16  2014 hg38.fa.gz
-rwxrwxrwx. 1 acressy3  gtperson  14K Sep 21 18:18 Project2-ReadMapping.ipynb


<div class="alert alert-block alert-warning">
    <h3>1.3 Find out about your reference genome file.</h3>
    <p>Does your reference fasta file have more information that just ACGT nucleotides? Please check if your fastq sequence have Ns or small/upper case ACGT nnucleotides.</p>
</div>

In [3]:
#Just glance over the fasta file. There is no need to run a script for this. 
#Write in one sentence if you see more than just upper case ACGT nucleotides.

In [3]:
!mkdir -p ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment

In [5]:
!gunzip ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.fa.gz

In [7]:
!ls -lh ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/
!head ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.fa
!head ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.fa -n20000 | tail

total 3.1G
-rw-r--r--. 1 soliver35 gtperson 3.1G Jan 16  2014 hg38.fa
>chr1
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN
AGGAGAGGTCGGTGCCGGGTCCCGGGGGTcccgcgccctccccccgcctc
caagccgccgccgcccgcgccTCACCCGTCACCTGCACGCGACGCAGGCT
CCGCAGGTGTCTCACGGTCATCTCCAGGATGTCCGCCTTCTCCAGCTTCG
AGTGGCGGGAGCTCTGGGGGCGGGGATAGGCGGGAGGTCCAGGTCAGCTG
CGACCCAGACTCCGGGTCTCGGGCCTTCGCCCCCGACTTACCTCTTTTCT
GAGGGCGTCCAGGATGAGGGTTTTGAGCTGAGCGAGGCTCTCGTTAATAC
GCGCTCGGCGCCGCTTCTCCATGACCGGCTTGGAGGACTGCGGGTCGGGC
ACCGGCTGAGTCCCGCGTCCCTCCCGCCCCCCGGTCGCCCCCCTCACGCC
CGGCCGGGACCCCACCTTGCGGTGCTCGGCCGCGCTCCGGGGCTTGTCTG
GGGTCC

There are 9 lines (in jupyter notebook) of Ns and about about 1 line of lowercase nucleotides

<div class="alert alert-block alert-warning">
    <h3>1.4 Using 1-3 lines, explain what upper case ACGT, lower case ACGT, and NNNs mean in your reference fasta file.</h3>
    <p>Hint: check readme files.</p>
</div>

Upper case ACGT are the non-repeating sequences, and lower cases ACGT are repeating sequences that are soft masked. 
The base sequences that could not be identified are shown by NNNs (i.e. missing bases, gaps). 

<div class="alert alert-block alert-warning">
    <h3>1.5 Select a tool for read alignment.</h3>
    <p>We have demonstrated bwa and minimap2 in the class. If you want to use some other tool, reach out to one of the TAs to get that tool installed on PACE-ICE.</p>
</div>

In [4]:
!bwa index
#actually indexed below


Usage:   bwa index [options] <in.fasta>

Options: -a STR    BWT construction algorithm: bwtsw, is or rb2 [auto]
         -p STR    prefix of the index [same as fasta name]
         -b INT    block size for the bwtsw algorithm (effective with -a bwtsw) [10000000]
         -6        index files named as <in.fasta>.64.* instead of <in.fasta>.* 

         `-a div' do not work not for long genomes.



# **2. Read alignment**
### *Total Questions: 4*
### *Total Points: 10+15+5+20 = 50*


---

<div class="alert alert-block alert-warning">
    <h3>2.1 Steps before mapping</h3>
    <p>You already have a fasta file, but sometimes alignment tools do not accept fasta files directly. Please read your aligners documentation and run all the preliminary steps require to run the mapping command from your aligner.</p>
</div>

<div class="alert alert-block alert-info">
    <p>Preliminary commands go in the cells below. Please use appropriate comments and markdown texts to explain what your command is doing.</p>
</div>

In [ ]:
#Write your commands here. Insert more cells if necessary

In [2]:
!ls -lh ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment

total 8.4G
-rwxrwxrwx. 1 soliver35 gtperson 3.1G Jan 16  2014 hg38.fa
-rw-r--r--. 1 jkim3607  gtperson  21K Sep 28 14:50 hg38.fa.amb
-rw-r--r--. 1 jkim3607  gtperson  22K Sep 28 14:50 hg38.fa.ann
-rw-r--r--. 1 jkim3607  gtperson 3.0G Sep 28 14:50 hg38.fa.bwt
-rw-r--r--. 1 jkim3607  gtperson 766M Sep 28 14:50 hg38.fa.pac
-rw-r--r--. 1 jkim3607  gtperson 1.5G Sep 28 15:04 hg38.fa.sa
-rw-r--r--. 1 jkim3607  gtperson    0 Sep 24 15:50 hg38.mmi
-rw-r--r--. 1 jkim3607  gtperson    0 Sep 28 11:43 SRR393028.sam


In [6]:
!ls ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment

hg38.fa  hg38.fa.amb  hg38.fa.ann  hg38.fa.pac	hg38.mmi  SRR393028.sam


In [1]:
#check the output
!head ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sam
!wc ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sam

@SQ	SN:chr1	LN:248956422
@SQ	SN:chr10	LN:133797422
@SQ	SN:chr11	LN:135086622
@SQ	SN:chr11_KI270721v1_random	LN:100316
@SQ	SN:chr12	LN:133275309
@SQ	SN:chr13	LN:114364328
@SQ	SN:chr14	LN:107043718
@SQ	SN:chr14_GL000009v2_random	LN:201709
@SQ	SN:chr14_GL000225v1_random	LN:211173
@SQ	SN:chr14_KI270722v1_random	LN:194050
   4572750   73377281 1102823023 /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sam


In [ ]:
#Run "ls -lh" to show what additional files were created in addition to your original fasta files.

In [1]:
!ls -lh ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment

total 9.4G
-rwxrwxrwx. 1 soliver35 gtperson 3.1G Jan 16  2014 hg38.fa
-rw-r--r--. 1 jkim3607  gtperson  21K Sep 28 14:50 hg38.fa.amb
-rw-r--r--. 1 jkim3607  gtperson  22K Sep 28 14:50 hg38.fa.ann
-rw-r--r--. 1 jkim3607  gtperson 3.0G Sep 28 14:50 hg38.fa.bwt
-rw-r--r--. 1 jkim3607  gtperson 766M Sep 28 14:50 hg38.fa.pac
-rw-r--r--. 1 jkim3607  gtperson 1.5G Sep 28 15:04 hg38.fa.sa
-rw-r--r--. 1 jkim3607  gtperson    0 Sep 24 15:50 hg38.mmi
-rwxrwxrwx. 1 jkim3607  gtperson 1.1G Sep 28 19:28 SRR393028.sam


<div class="alert alert-block alert-warning">
    <h3>2.2 Run mapping.</h3>
    <p>This should be one or two lines of code but it's important to understand what the options are doing.</p>
</div>

In [ ]:
#Write you command below.

In [1]:
#index the file using bwa
!bwa index -a rb2 ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.fa

[bwa_index] Pack FASTA... 24.90 sec
[bwa_index] Construct BWT for the packed sequence...
[bwa_index] 5708.25 seconds elapse.
[bwa_index] Update BWT... 16.41 sec
[bwa_index] Pack forward-only FASTA... 14.95 sec
[bwa_index] Construct SA from BWT and Occ... 779.76 sec
[main] Version: 0.7.17-r1188
[main] CMD: bwa index -a rb2 /home/hice1/jkim3607/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.fa
[main] Real time: 6605.213 sec; CPU: 6544.299 sec


In [1]:
#align the file using bwa

!bwa mem -t 24 ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.fa ~/biol6150/ProjectSubmissions/Group8/Project1/Trimming/SRR393028_1.Trimmed.fastq ~/biol6150/ProjectSubmissions/Group8/Project1/Trimming/SRR393028_2.Trimmed.fastq > ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sam


[M::bwa_idx_load_from_disk] read 0 ALT contigs
[M::process] read 3583244 sequences (240000084 bp)...
[M::mem_pestat] # candidate unique pairs for (FF, FR, RF, RR): (22, 1345179, 6, 8)
[M::mem_pestat] analyzing insert size distribution for orientation FF...
[M::mem_pestat] (25, 50, 75) percentile: (73, 115, 139)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 271)
[M::mem_pestat] mean and std.dev: (109.23, 49.23)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 337)
[M::mem_pestat] analyzing insert size distribution for orientation FR...
[M::process] read 988624 sequences (66206370 bp)...
[M::mem_pestat] (25, 50, 75) percentile: (104, 135, 174)
[M::mem_pestat] low and high boundaries for computing mean and std.dev: (1, 314)
[M::mem_pestat] mean and std.dev: (142.33, 50.39)
[M::mem_pestat] low and high boundaries for proper pairs: (1, 384)
[M::mem_pestat] skip orientation RF as there are not enough pairs
[M::mem_pestat] skip orientation RR as ther

In [1]:
#index
!minimap2 -d ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.mmi ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.fa
!ls -lh ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/

[M::mm_idx_gen::60.096*1.62] collected minimizers
[M::mm_idx_gen::75.676*1.90] sorted minimizers
[M::main::75.676*1.90] loaded/built the index for 455 target sequence(s)
[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 455
[M::mm_idx_stat::76.892*1.88] distinct minimizers: 100202295 (37.96% are singletons); average occurrences: 5.732; average spacing: 5.587; total length: 3209286105
[M::main] Version: 2.26-r1175
[M::main] CMD: minimap2 -d /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.mmi /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.fa
[M::main] Real time: 77.056 sec; CPU: 145.093 sec; Peak RSS: 11.974 GB
total 9.4G
-rwxrwxrwx. 1 soliver35 gtperson 3.1G Jan 16  2014 hg38.fa
-rw-r--r--. 1 jkim3607  gtperson  21K Sep 28 14:50 hg38.fa.amb
-rw-r--r--. 1 jkim3607  gtperson  22K Sep 28 14:50 hg38.fa.ann
-rw-r--r--. 1 jkim3607  gtperson 3.0G Sep 28 14:50 hg38.fa.bwt
-rw-r--r--. 1 jkim3607  gtperson 766M Sep 

In [ ]:
#Add one extra cell as required and explain each argument/flag that was given to the mapping command.

In [1]:
#Align
!minimap2 -c -a ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.mmi ~/biol6150/ProjectSubmissions/Group8/Project1/Trimming/SRR393028_1.Trimmed.fastq ~/biol6150/ProjectSubmissions/Group8/Project1/Trimming/SRR393028_2.Trimmed.fastq > ~/scratch/ReadAlignment/SRR393028.Minimap.sam  

[M::mm_idx_gen::0.004*0.91] collected minimizers
[M::mm_idx_gen::0.005*0.75] sorted minimizers
[M::main::0.020*0.18] loaded/built the index for 0 target sequence(s)
[M::mm_mapopt_update::0.020*0.19] mid_occ = 898015
[M::mm_idx_stat] kmer size: 15; skip: 10; is_hpc: 0; #seq: 0
[M::mm_idx_stat::0.020*0.19] distinct minimizers: 0 (-nan% are singletons); average occurrences: -nan; average spacing: -nan; total length: 0
[M::worker_pipeline::8.217*0.81] mapped 2285934 sequences
[M::worker_pipeline::15.996*0.83] mapped 2285934 sequences
[M::main] Version: 2.26-r1175
[M::main] CMD: minimap2 -c -a /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/hg38.mmi /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project1/Trimming/SRR393028_1.Trimmed.fastq /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project1/Trimming/SRR393028_2.Trimmed.fastq
[M::main] Real time: 15.996 sec; CPU: 13.269 sec; Peak RSS: 0.648 GB


<div class="alert alert-block alert-warning">
    <h3>2.3 Post alignment evaluation.</h3>
    <p>How many lines does your SAM file have?</p>
</div>

In [ ]:
#Just one line for this one.

In [2]:
!wc -l ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sam

4572750 /home/hice1/acressy3/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sam


We have 4572750 lines in our SAM file.

<div class="alert alert-block alert-warning">
    <h3>2.4 Look at the 7th read that maps successfully in your SAM file and answer the following questions. (10+10+10+10 = 40Points)</h3>
    <p>1. What chromosome does it align to?</p>
    <p>2. What is the CIGAR string for this read? Explain the CIGAR string in detail.</p>
    <p>3. Does this read have a paired read which also aligns successfully?</p>
    <p>4. If the paired read aligns successfully, what does the CIGAR string for this read convey?</p>
</div>

In [9]:
!head -500 ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sam

@SQ	SN:chr1	LN:248956422
@SQ	SN:chr10	LN:133797422
@SQ	SN:chr11	LN:135086622
@SQ	SN:chr11_KI270721v1_random	LN:100316
@SQ	SN:chr12	LN:133275309
@SQ	SN:chr13	LN:114364328
@SQ	SN:chr14	LN:107043718
@SQ	SN:chr14_GL000009v2_random	LN:201709
@SQ	SN:chr14_GL000225v1_random	LN:211173
@SQ	SN:chr14_KI270722v1_random	LN:194050
@SQ	SN:chr14_GL000194v1_random	LN:191469
@SQ	SN:chr14_KI270723v1_random	LN:38115
@SQ	SN:chr14_KI270724v1_random	LN:39555
@SQ	SN:chr14_KI270725v1_random	LN:172810
@SQ	SN:chr14_KI270726v1_random	LN:43739
@SQ	SN:chr15	LN:101991189
@SQ	SN:chr15_KI270727v1_random	LN:448248
@SQ	SN:chr16	LN:90338345
@SQ	SN:chr16_KI270728v1_random	LN:1872759
@SQ	SN:chr17	LN:83257441
@SQ	SN:chr17_GL000205v2_random	LN:185591
@SQ	SN:chr17_KI270729v1_random	LN:280839
@SQ	SN:chr17_KI270730v1_random	LN:112551
@SQ	SN:chr18	LN:80373285
@SQ	SN:chr19	LN:58617616
@SQ	SN:chr1_KI270706v1_random	LN:175055
@SQ	SN:chr1_KI270707v1_random	LN:32032
@SQ	SN:chr1_KI270708v1_random	LN:127682
@SQ	SN:chr1_KI270709v1_rando

1. The seventh read (SRR393028.7) aligns to to chromosome 1.

2. The SAM file spits out 11 different fields that correspond to different descriptions. The 6th field is the CIGAR string, usually in the format of a number followed by a capital letter (or multiple number-letter combos). The CIGAR string for this read is 67M =. It means there are 67 matches because M stands for match or mistmatch, but followed by an equals sign, it means there is a match. If followed by an X, this would mean there is a mismatch. Some other letters include I=insertion, D=deletion, N=skipped region, S=soft clipping, H=hard clipping, P=padding, X=sequence mismatch, and == sequence match. 

3. This read has a paired read that also aligns successfully. We know this because there is a SAM flag value of 99 and 147. A SAM flag value of 99 means that the read is properly paired, has a mate reverse strand, and is first in pair. A SAM flag value of 147 means the same thing but that it is second in a pair.

4. Since the paired reads aligned successfully, the CIGAR string conveys that 67 base pairs matched successfully. 

# **3. Read alignment evaluation**
### *Total Questions: 5*
### *Total Points: 2+5+10+3+10 = 30*


---

<div class="alert alert-block alert-warning">
    <h3>3.1 Samtools is a classic bioinformatics tools.</h3>
    <p>Check if you have samtools installed on your environment.</p>
</div>

In [3]:
#Just run the "samtools" command here and make sure it does not say "samtools not found".
!samtools


Program: samtools (Tools for alignments in the SAM format)
Version: 1.17 (using htslib 1.17)

Usage:   samtools <command> [options]

Commands:
  -- Indexing
     dict           create a sequence dictionary file
     faidx          index/extract FASTA
     fqidx          index/extract FASTQ
     index          index alignment

  -- Editing
     calmd          recalculate MD/NM tags and '=' bases
     fixmate        fix mate information
     reheader       replace BAM header
     targetcut      cut fosmid regions (for fosmid pool only)
     addreplacerg   adds or replaces RG tags
     markdup        mark duplicates
     ampliconclip   clip oligos from the end of reads

  -- File operations
     collate        shuffle and group alignments by name
     cat            concatenate BAMs
     consensus      produce a consensus Pileup/FASTA/FASTQ
     merge          merge sorted alignments
     mpileup        multi-way pileup
     sort           sort alignment file
     split          splits a

<div class="alert alert-block alert-warning">
    <h3>3.2 Is your SAM file sorted by coordinates?</h3>
    <p>SAM or BAM (which we will create later) can be sorted by the read name or read coordinate. Look at your SAM file and see if you can run a Samtools command to find out if your file is sorted. (Hint: Google your question)</p>
</div>

In [11]:
#Tell us if your SAM file is sorted.

Our SAM file is not sorted.

In [12]:
#Tell us how you reached your answer and clerly write the commands used in additional cell with descriptive comments.

In [12]:
!tail ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sam

SRR393028.2305516	83	chr12	55839646	60	67M	=	55839576	-137	GGTTCACCCCCCGGTAGGTCCCCTCAGTCCCGAACTCGTCTTCCTCGAAGTTCACACTGCCCAGCTC	???????????????????????????????????????????????????????????????????	NM:i:4	MD:Z:10A19A12G7G15	MC:Z:1S35M31S	AS:i:47	XS:i:0
SRR393028.2305516	163	chr12	55839576	44	1S35M31S	=	55839646	137	TAGGGCCTGGGATTATCGGGACTGCCAAAGCCCCAGTATGAGGGCCGCTTCTGGGGCTGCAATGATG	???????????????????????????????????????????????????????????????????	NM:i:3	MD:Z:11A8G4C9	MC:Z:67M	AS:i:20	XS:i:0
SRR393028.2305517	83	chr11	71082729	60	67M	=	71082685	-111	CTGGAGCCGGAGGAGAGGATAAGTCACTGTGGCCCGAAACTATTTCTTTAAGGCGGACACTCTTCAT	???????????????????????????????????????????????????????????????????	NM:i:2	MD:Z:7A11G47	MC:Z:67M	AS:i:57	XS:i:19
SRR393028.2305517	163	chr11	71082685	60	67M	=	71082729	111	GTCAAGAGCTGAGGGCATCCGGCCAGGGATTGGCAGGGGGCAGCCTGGAGCAGGATGAGAGGAGAAG	???????????????????????????????????????????????????????????????????	NM:i:3	MD:Z:0C29G24G11	MC:Z:67M	AS:i:56	XS:i:22
SRR393028.2305518	77	*

<div class="alert alert-block alert-warning">
    <h3>3.3 Sort your SAM file?</h3>
    <p>Sort your SAM file by genomic coordinates,</p>
    <p>When you run this command, what is the output format of your result file?</p>
    <h4>Tweak samtools and change your sorted SAM file's output to SAM. (Hint: You should be able to read the file using simple cat or less)</h4>
</div>

Our result SAM file is sorted by coordinates. We know this because SO in the @HD line means sorting order.

In [13]:
#Run the sort command here.
!samtools sort ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sam -o ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sorted.sam

[bam_sort_core] merging from 1 files and 1 in-memory blocks...


In [ ]:
#Head your sorted SAM file here. Show the first 5 lines.

In [15]:
!head -5 ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sorted.sam

@HD	VN:1.6	SO:coordinate
@SQ	SN:chr1	LN:248956422
@SQ	SN:chr10	LN:133797422
@SQ	SN:chr11	LN:135086622
@SQ	SN:chr11_KI270721v1_random	LN:100316


In [ ]:
#Write any other commands used here as well.

<div class="alert alert-block alert-warning">
    <h3>3.4 Convert your SAM file to a BAM file.</h3>
    <p>This should be a single one line command.</p>
</div>

In [16]:
!samtools view -S -b ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sorted.sam > ~/biol6150//ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sorted.bam

In [18]:
#Write the command run "ls -lh" to show the size and location of the file.
!ls -lh ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment

total 11G
-rwxrwxrwx. 1 soliver35 gtperson 3.1G Jan 16  2014 hg38.fa
-rw-r--r--. 1 jkim3607  gtperson  21K Sep 28 14:50 hg38.fa.amb
-rw-r--r--. 1 jkim3607  gtperson  22K Sep 28 14:50 hg38.fa.ann
-rw-r--r--. 1 jkim3607  gtperson 3.0G Sep 28 14:50 hg38.fa.bwt
-rw-r--r--. 1 jkim3607  gtperson 766M Sep 28 14:50 hg38.fa.pac
-rw-r--r--. 1 jkim3607  gtperson 1.5G Sep 28 15:04 hg38.fa.sa
-rw-r--r--. 1 jkim3607  gtperson    0 Sep 24 15:50 hg38.mmi
-rwxrwxrwx. 1 jkim3607  gtperson 1.1G Sep 28 19:28 SRR393028.sam
-rw-r--r--. 1 acressy3  gtperson 141M Sep 29 20:37 SRR393028.sorted.bam
-rwxrwxrwx. 1 acressy3  gtperson 1.1G Sep 29 20:27 SRR393028.sorted.sam


<div class="alert alert-block alert-warning">
    <h3>3.5 Statistics of your SAM file.</h3>
    <p>1.How many total raw sequences does your SAM file have?</p>
    <p>2.How many total reads are mapped in your SAM file?</p>
</div>

In [1]:
#Show your commands and answers in additional cells as required.
#counting total raw sequences in SAM file
!samtools view -c ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sorted.sam

4572294


In [1]:
#double checking the raw sequence count
!grep -v '^@' ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sorted.sam | wc

4572294 73375905 1102806463


In [1]:
#counting total reads mapped in SAM file
!samtools flagstat ~/biol6150/ProjectSubmissions/Group8/Project2/ReadAlignment/SRR393028.sorted.sam

4572294 + 0 in total (QC-passed reads + QC-failed reads)
4571868 + 0 primary
0 + 0 secondary
426 + 0 supplementary
0 + 0 duplicates
0 + 0 primary duplicates
4455191 + 0 mapped (97.44% : N/A)
4454765 + 0 primary mapped (97.44% : N/A)
4571868 + 0 paired in sequencing
2285934 + 0 read1
2285934 + 0 read2
4358108 + 0 properly paired (95.32% : N/A)
4405492 + 0 with itself and mate mapped
49273 + 0 singletons (1.08% : N/A)
30440 + 0 with mate mapped to a different chr
24231 + 0 with mate mapped to a different chr (mapQ>=5)


In [17]:
#You can parse the SAM file manually, or figure out a faster way to do this.

1. The SAM file has 4572294 raw sequences, as detemined with the samtools view -c command, which counts the number of reads in the file. 
2. The SAM file has 4455191 total reads mapped (97.44%). This was determined using samtools flagstat, which returns statistics for the SAM file. The number of reads mapped can be determined from the line containing "+ 0 mapped."